### ISSUE TITLE
support for hardened/musl profile?

### ISSUE THREAD

**Issue by sarnold** [Obsidian-StudiosInc/os-xtoo#65]
I have a fork with some packages I needed for testing (excuse the two honking bin packages, I just ran out of time and had to stop chasing old dependencies).  BUT the only musl option I found was the upstream openjdk-12-ea+20_linux-x64-musl package.  Did you want to look at my fork?  (sorry, no spoons)

https://github.com/teamorchard/os-xtoo/tree/musl-test

There is an icedtea-3.8.0 source package in the musl overlay but I couldn't find anything to build it with on musl :/

**Comment by wltjr** [comment]
Thanks for mentioning the fork. I am open to supporting musl in addition to glibc. I think most can be handled in eclasses, like arch keywords. Trying to centralize stuff so various changes do not require touching actual ebuilds. Though the native packages may need some stuff. Not to many Java packages that build native components, so touching those is not that bad.

I am not a big icedtea fan, first off it should be called openjdk not renamed due to some build harness. But we may have another option. Oracle builds [Java 12 under Alpine Linux](http://jdk.java.net/12/) which is musl based. Not sure if you are aware of that. Maybe we can use that to move forward till there are other options.

I need to check out the openjdk from source build in Gentoo's main repo someone contributed. It is not using Icedtea at all. I think that may build under musl. That is one of the last pieces to build from source, the actual JDK. But given the bootstrapping aspect, its not be a high priority item. I know there is a difference between running a pre-built binary and building from source via bootstrap. Just tend to like as few JDKs on my system as possible, if not just 1.

Anyway thanks for mentioning your fork. I will check it out ASAP, but bit busy so likely be a week or two. I am happy to work with you to merge in any support for musl or modify this repo so it has full support for such. Either way I am flexible all around.

**Comment by sarnold** [comment]
You mean there's also an Oracle bin release too?  (openjdk-12-ea+20 is also the alpine thing i think)

I'm not really the java guy, and I really only have a requirement to support a (java-based) test harness, which means I just ran into musl vs. java and I got pointed here.  So mainly I just need a working jdk for musl/hardened/selinux profile with enough deps packaged so it's (theoretically) possible to use stuff like SikuliX for testing.

**Comment by wltjr** [comment]
@sarnold the openjdk 12 is from Oracle just not officially oracle branded. Not sure if your able to run the one for Alpine Linux.

I need to get back to more work on the JDK aspect for other archs and musl vs glibc. I have mostly focused on updating dependencies and keeping Java packages in general current. Not sure if you have much use of that stuff. Or if you can find use of the core in my overlay. Hopefully at min I can get the base stuff working for musl and then can see about any specific packages.

It maybe semi trivial, if all that is needed is [changing deps](https://github.com/Obsidian-StudiosInc/os-xtoo/blob/master/dev-java/openjdk-bin/openjdk-bin-9999.ebuild#L58) in the JVM package. Not seeing anything relating to glibc in eclasses. I think that maybe the only dependency. Which could likely be switched to some conditional on musl or glibc, rather than just glibc.

**Comment by wltjr** [comment]
@sarnold now that jem has been fixed for musl. I think all that is needed now is to modify the jdk ebuilds to bring in the right sources when running under musl. Which at the moment can likely only be done with oracles openjdk 12 releases, since they build those under Alpine Linux using musl. I assume that is the binary you used for your purposes. If you found another please let me know, thanks!

**Comment by sarnold** [comment]
Nope, the one apline early access bin package is the only (newer) musl-built jdk I could find.  The main tree only has very minimal (older) jdk support, and I could not bootstrap from one to the other.  Also we're doing this on an experimental profile, so now seems like the time to get the new java stuff in if we can (I just don't have much "extra" time right now).  Note all that stuff on the musl-test branch builds/installs on musl profile using the openjdk-12 musl package; I think mainly it needs that small patch for jem and a few package tweaks here and there...

**Comment by wltjr** [comment]
Ok good to know thanks!

I am not sure if the openjdk from source ebuild in tree could be used someday to build one locally supporting musl. Though I think for performance reasons you may have to wait till Oracle does a release of 12, hopefully with a musl built binary available. It seems that Oracle will only provide Hotspot under their licensed versions, openjdk-11 uses zero not hotspot as the JVM. Oracle JDK 11 uses hotspot. Something I was not previously aware of, that may effect musl jdk options. That will remain for 12 and onward. Which will likely keep some using Oracle binaries vs any from source built alternative.

**Comment by sarnold** [comment]
The musl source patches should be pretty small, mostly header fixes I would think.  Doesn't alpine make their source an/or patches available?

**Comment by wltjr** [comment]
I have to look into the Alpine situation, but I think it is just one of luck and chance. Oracle uses that distribution to build a jdk supporting musl. Not many commits reference musl but seems they are putting some into openjdk sources, at least this [one commit](https://github.com/unofficial-openjdk/openjdk/commit/b4cbe3d9c6f2126a8a4908129d09ff3854e7ff9d) references alpine/musl.

I think Oracle is just making the binaries for Alpine Linux maybe. I believe Oracle is the one producing the early access builds. Though if all the musl changes were in openjdk, then I think there would be some alternative builds under musl available. Those alternatives do not seem to exist. Presently as you confirmed the only jdk supporting musl is the binary one from Oracle, thus I assume they are the ones making those.

I think it just happens they chose to make them on Alpine vs some other distro. Likely because Alpine is fully built around musl. I think only Gentoo can do a 100% musl system, like you are working on, but its from source not binary, which could lead to variances not experienced under a binary distro. I think Alpine is the only binary distro that supports musl only. Others are still glibc based and have musl built packages like on Ubuntu. That is more a multi libc distro than one specific to glibc, musl, or other ulibc, etc. Given that, Alpine was likely the only distro Oracle could use and build under, thus making those packages for Alpine. Though Oracle may have some grander plan for Alpine. I haven't ever seen them target a specific Linux distro like that before. Maybe something Oracle will use for their Linux offerings and move away from RHEL.
